### PR TITLE
Add cookie consent functions

### LINKED ISSUE #1614: Build cookie consent functions in the design system
## What
Add the cookie consent mechanism to the design system. This will consist of a set of functions to:

- accept and reject cookies
- save consent in a cookie, with an expiry of 1 year
- delete cookies when a user changes their preference to reject
- look up a user's consent
- be able to map a cookie name to it's category, e.g: `_ga` -> `analytics`

**Example in GOV.UK**: https://github.com/alphagov/govuk_publishing_components/blob/master/app/assets/javascripts/govuk_publishing_components/lib/cookie-functions.js

Part of this work will be deciding what the consent cookie looks like, e.g: it's name; what the value will look like

We should be ok to merge these changes as we're adding a set of functions without using them, so they won't have any effect on users yet.

## Why
We will need this set of functions to hook into when we add the cookie banner and cookie preferences form.

## Who needs to know about this
Developers

## Done when

- [ ] We know what the consent cookie will look like
- [ ] We have functions for the main cookie banner/form behaviour, e.g: accepting and rejecting cookies; deleting cookies (see list above)
- [ ] Tests for all of the cookie consent functions

### PR DESCRIPTION
**Pull request by lfdebrux**
Closes #1614.

This PR adds copies of the [cookie consent functions from digitalmarketplace-govuk-frontend](https://github.com/alphagov/digitalmarketplace-govuk-frontend/tree/d24e7f1dc6e94fb9f173aebf46aed5f1051af59c/src/digitalmarketplace/helpers/cookie), and refactors them slightly.

---

We wanted a cookie module to provide the following functionality:

- accept and reject cookies
- save consent in a cookie, with an expiry of 1 year
- delete cookies when a user changes their preference to reject
- look up a user's consent
- be able to map a cookie name to it's category, e.g: `_ga` -> `analytics`

I reviewed existing code in other alphagov projects and found the Digital Marketplace GOV.UK Frontend code fit all of our needs and was the best fit for our existing codebase. See issue #1615 for more details of the analysis done. I also ran the analysis past @vanitabarrett for a sense check.

---

This commit copies `cookie-functions.js` and `cookie-functions.test.js` to `src/javascripts` where it can be used by future code for the cookie banner and cookie page. Currently the code is unused and is not sent to browsers.

[1]: https://github.com/alphagov/digitalmarketplace-govuk-frontend/tree/d24e7f1dc6e94fb9f173aebf46aed5f1051af59c/src/digitalmarketplace/helpers/cookie

### CODE REVIEW

**Comment by lfdebrux** [comment]
Currently the cookie that holds the users cookie consent preferences is called `cookie_policy`; I'm not sure if we're happy with that name or not? Do we have a naming convention for cookies?

File: src/javascripts/cookie-functions.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  setCookie('dm_cookies_policy', JSON.stringify(cookieConsent), { days: 365 })
+}
+
+function checkConsentCookieCategory (cookieName, cookieCategory) {
```
Comment (vanitabarrett) [inline src/javascripts/cookie-functions.js:110]: I know this isn't your change, I wonder if we can rename this function to make it a bit clearer what it does? Maybe something like `checkUserConsentForCategory` . What do you think?

**Review by vanitabarrett — CHANGES_REQUESTED** [review]

**Comment by vanitabarrett** [comment]
In terms of the cookie name, perhaps we want to make it a bit clearer that it's linked to the design system? Especially if we add sub-domain consent later. Maybe `ds_cookie_policy` or `design_system_cookie_policy` 🤷‍♀️

File: src/javascripts/cookie-functions.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
+// used by the cookie banner component
+const DEFAULT_COOKIE_CONSENT = {
+  analytics: false
+}
+
+const COOKIE_CATEGORIES = {
+  _ga: 'analytics',
+  _gid: 'analytics',
+  _gat_govuk_shared: 'analytics'
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:13]: Ah, interesting

File: src/javascripts/cookie-functions.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    } catch (err) {
+      return null
+    }
+
+    if (typeof consentCookieObj !== 'object' && consentCookieObj !== null) {
+      consentCookieObj = JSON.parse(consentCookieObj)
+    }
+  } else {
+    return null
+  }
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:78]: Interesting, the tests pass without those two lines!

File: src/javascripts/cookie-functions.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
+// used by the cookie banner component
+const DEFAULT_COOKIE_CONSENT = {
+  analytics: false
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:7]: Hmm, thinking about it, adding `essential: true` suggests that a user could opt-out of essential cookies. Is that correct? Are there any circumstances where there would be cookies a user couldn't opt out of?

File: src/javascripts/cookie-functions.js (hunk 1 above)
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:110]: How about `checkConsentForCookieCategory`?

We might then also want to rename the function below to `checkConsentForCookie`, what do you think?

File: src/javascripts/cookie-functions.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  setCookie('hobnob', 'tasty', { days: 30 })
+
+  Reading a cookie:
+  getCookie('hobnob')
+
+  Deleting a cookie:
+  Cookie('hobnob', null)
+*/
+
+function Cookie (name, value, options) {
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:46]: I think exporting `Cookie` and not `(set|get)Cookie` makes more sense to me.

File: src/javascripts/cookie-functions.test.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
+/**
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.test.js:1]: Yeah, I mention this in the commit message, I looked at the other tests in `__tests__` and noticed they were all using Puppeteer and were testing the integration of the JavaScript with the page. These tests are unit tests that test the JavaScript standing alone. I considered that we might later want to add integration tests for the cookie functions, or unit tests for the other JavaScripts, so I put these unit tests in a different place to show that they are different.

[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.test.js:1]: I'm happy to move these new tests to `__tests__` though if you think that's better.

File: src/javascripts/cookie-functions.js (hunk 1 above)
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:110]: Looking more closely at what these functions do and are used for, maybe `hasUserConsentedToCookie` and `hasUserConstentedToCookieCategory` would be better? Or is that too wordy?

File: src/javascripts/cookie-functions.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  setCookie('dm_cookies_policy', JSON.stringify(cookieConsent), { days: 365 })
+}
+
+function checkConsentCookieCategory (cookieName, cookieCategory) {
+  var currentConsentCookie = getConsentCookie()
+
+  // If the consent cookie doesn't exist, but the cookie is in our known list, return true
+  if (!currentConsentCookie && COOKIE_CATEGORIES[cookieName]) {
+    return true
+  }
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:109]: Ah. Good spot. Thanks.

File: src/javascripts/cookie-functions.js (hunk 1 above)
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js]: Another option 😅 : `allowSettingCookie`, `allowSettingCookieInCategory`?

File: src/javascripts/cookie-functions.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:109]: So is the correct behaviour here be to check against `DEFAULT_COOKIE_CONSENT`? Or should it be more restrictive even?

File: src/javascripts/cookie-functions.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+ * ================
+ *
+ * Used by the cookie banner component and cookies page pattern.
+ *
+ * Includes function `Cookie()` for getting, setting, and deleting cookies, and
+ * functions to manage the users' consent to cookies.
+ */
+
+/* Name of the cookie to save users cookie preferences to. */
+const CONSENT_COOKIE_NAME = 'cookies_policy'
```
Comment (vanitabarrett) [inline src/javascripts/cookie-functions.js:12]: I don't think we can use `const`?

File: src/javascripts/cookie-functions.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (vanitabarrett) [reply src/javascripts/cookie-functions.js:16]: True - a user won't be able to opt out of essential cookies. Essential cookies are only those which are essential to the running of the site (either the site breaks or a large portion of it doesn't work without cookies).

File: src/javascripts/cookie-functions.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (vanitabarrett) [reply src/javascripts/cookie-functions.js:109]: I think checking against `DEFAULT_COOKIE_CONSENT` should work...

File: src/javascripts/cookie-functions.js (hunk 8 above)
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:12]: Oh really? :( How come?

File: src/javascripts/cookie-functions.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:16]: Got it. I'm guessing the cookie policy cookie is essential then? As the cookie banner won't work without it.

[the remark this answers is not in the document]
Comment (vanitabarrett) [reply src/javascripts/cookie-functions.js:16]: Yep, I think that's the only essential cookie we have

File: src/javascripts/cookie-functions.js (hunk 8 above)
Comment (vanitabarrett) [reply src/javascripts/cookie-functions.js:12]: I don't think we transpile these JS files. I think this PR passes because the code in this file isn't being called or imported anywhere. Once you do that ([example](https://github.com/alphagov/govuk-design-system/compare/ldeb-add-cookie-functions...add-gtm-snippet)), you get errors on build/start from UglifyJS:

```
Error: UglifyJS Error: {"message":"Unexpected token: keyword «const»","filename":"javascripts/application.js","line":9048,"col":1,"pos":309281}
```

Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:12]: Oh no 😢 Is there alternative to `const` that is good to use?

Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:12]: Looks like there is no way to transpile `const` anyway.

https://stackoverflow.com/questions/57912755/javascript-es5-transpile-for-es6-const#comment102245785_57912755

I will just substitute with `var`.

Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:12]: Do you know if upgrading `uglify-js` allows us to use const? The release we are using is 2 years old now.

File: src/javascripts/cookie-functions.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (36degrees) [reply src/javascripts/cookie-functions.test.js:1]: My only reservation is that I'm pretty these will end up being copied into the 'built' site at the minute, which I think is why the other tests are outside of `/src/`. That feels a bit weird to me, but maybe it's OK.

File: src/javascripts/cookie-functions.js (hunk 8 above)
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js:12]: Bumping `metalsmith-uglify` in main and rebasing this PR means we can use `const` now 🎉

File: src/javascripts/cookie-functions.js (hunk 1 above)
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js]: Went with `userAllowsCookie` and `userAllowsCookieCategory`.

File: src/javascripts/cookie-functions.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js]: Okay, I've made that change in the latest commits.

File: src/javascripts/cookie-functions.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.js]: I went with the approach of not adding `essential: true` to the default cookies, but adding an explicit line to the checking logic that checks whether an cookie's category is `essential`. Have a look at the commit message in c5ed9bd7bb032025e1a7e09e72042783063472bf for more details.

File: src/javascripts/cookie-functions.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.test.js:1]: Ah, well, that's a good reason to put the test in `__tests__` then!

[the remark this answers is not in the document]
Comment (lfdebrux) [reply src/javascripts/cookie-functions.test.js]: Moved it now.

File: __tests__/cookie-functions.test.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    })
+
+    it('returns consent cookie object if present', async () => {
+      document.cookie = 'design_system_cookies_policy={"analytics":false}'
+
+      expect(CookieHelpers.getConsentCookie()).toEqual({ analytics: false })
+    })
+  })
+
+  describe('setConsentCookie', () => {
```
Comment (vanitabarrett) [inline __tests__/cookie-functions.test.js:85]: Do we need something which tests that a consent cookie is set with the default values if no values are passed in?

Comment (lfdebrux) [reply __tests__/cookie-functions.test.js:85]: Good idea

Comment (lfdebrux) [reply __tests__/cookie-functions.test.js:85]: Done

**Review by vanitabarrett — APPROVED** [review]
Great work on this @lfdebrux ! 🎉
